### ISSUE TITLE
[FEATURE] Support LH priorityClass

### ISSUE THREAD

**Issue by FrankYang0529** [harvester/harvester#3756]
**Is your feature request related to a problem? Please describe.**
LH is a critical service. We don't want that it is the first to be evicted when eviction happens, so we would like to set priority class to LH workloads.

**Describe the solution you'd like**
We must detach all volumes when updating the priority class in old clusters. We have to update the LH helm chart to support old and new values in the same template.

**Describe alternatives you've considered**
If we want to support the value in old clusters, we should introduce the state machine like storage network.

**Additional context**

**Comment by bk201** [comment]
We need a doc to describe what to do in the upgrade path. Reviewing: https://github.com/harvester/harvesterhci.io/pull/37

**Comment by TachunLin** [comment]
### Test Plans

#### New Cluster 

- [x] 1. When the cluster is ready, check Longhorn system-managed components have priority class set.

    <details><summary>Test Result $~~$ $\color{green}{\textsf{PASS}}$ </summary>
    
        
      node1:~ # kubectl get priorityclass system-cluster-critical
      NAME                      VALUE        GLOBAL-DEFAULT   AGE
      system-cluster-critical   2000000000   false            173m
        
          
    </details> 

- [x] 2. Verify all system-managed components' pods have the correct priority, including 


    <details><summary>Test Result $~~$ $\color{green}{\textsf{PASS}}$ </summary> 
    
        node1:~ # kubectl get pods -n longhorn-system -o custom-columns="Name":metadata.name,"Priority":.spec.priority
        Name                                                  Priority
        csi-attacher-6d46d6df-6wdfk                           2000000000
        csi-attacher-6d46d6df-d9sg6                           2000000000
        csi-attacher-6d46d6df-fnv5w                           2000000000
        csi-provisioner-789fb46d84-224k5                      2000000000
        csi-provisioner-789fb46d84-q7b2b                      2000000000
        csi-provisioner-789fb46d84-rdclf                      2000000000
        csi-resizer-7b4c499ff9-df79r                          2000000000
        csi-resizer-7b4c499ff9-dv9p9                          2000000000
        csi-resizer-7b4c499ff9-vfpmm                          2000000000
        csi-snapshotter-865d9b6f88-76nrt                      2000000000
        csi-snapshotter-865d9b6f88-rzkxj                      2000000000
        csi-snapshotter-865d9b6f88-x4x95                      2000000000
        engine-image-ei-1d169b76-5mw5m                        2000000000
        engine-image-ei-1d169b76-8hbzf                        2000000000
        engine-image-ei-1d169b76-jbrsm                        2000000000
        instance-manager-e-203e966d715dbc1dc2cdd4fe7751f2a1   2000000000
        instance-manager-e-3dedec811d6f75d1ab229b5c8e988f76   2000000000
        instance-manager-e-98d0465194e538e2ef927749661973ba   2000000000
        instance-manager-r-203e966d715dbc1dc2cdd4fe7751f2a1   2000000000
        instance-manager-r-3dedec811d6f75d1ab229b5c8e988f76   2000000000
        instance-manager-r-98d0465194e538e2ef927749661973ba   2000000000
        longhorn-admission-webhook-86578b785d-79tfp           0
        longhorn-admission-webhook-86578b785d-vngnd           0
        longhorn-conversion-webhook-67b57c79cb-5rtk6          0
        longhorn-conversion-webhook-67b57c79cb-ppmnx          0
        longhorn-csi-plugin-ct22w                             2000000000
        longhorn-csi-plugin-qzvmh                             2000000000
        longhorn-csi-plugin-svb69                             2000000000
        longhorn-driver-deployer-76b8ddc678-xfccj             0
        longhorn-loop-device-cleaner-5s4kb                    0
        longhorn-loop-device-cleaner-mf88x                    0
        longhorn-loop-device-cleaner-qlw6g                    0
        longhorn-manager-j4p84                                0
        longhorn-manager-mrfx7                                0
        longhorn-manager-rjtzb                                0
        longhorn-recovery-backend-5776964c79-djv9l            0
        longhorn-recovery-backend-5776964c79-gm6lg            0
        longhorn-ui-7464fdc77c-2bn7d                          0
        longhorn-ui-7464fdc77c-4gjm9                          0      
              
    </details> 


    - [x] - csi-attacher
    - [x] - csi-provisioner
    - [x] - csi-resizer
    - [x] - csi-snapshotter
    - [x] - engine-image-ei
    - [x] - instance-manager-e
    - [x] - instance-manager-r
    - [x] - longhorn-csi-plugin




#### Upgrade from v1.1.2 to v1.2.0-rc
Refer to https://github.com/harvester/harvesterhci.io/pull/37 to set up PriorityClass

- [x]  1. After upgrade, manually set the `PriorityClass` with "system-cluster-critical" to Longhorn system-managed components pods 

    <details><summary>Test Result $~~$ $\color{green}{\textsf{PASS}}$ </summary>
    
     - We can follow the steps in [document](https://github.com/harvester/harvesterhci.io/blob/fb8a5558f6e89a8ae9ca5f7485ad8b18c7793d81/kb/2023-07-25/configure_priority_class_longhorn.md) to configure the LH priority class after upgrade 
        
         
    </details> 

- [x] 2. Verify all system-managed components' pods have the correct priority, including 

    <details><summary>Test Result $~~$ $\color{green}{\textsf{PASS}}$ </summary>
    
      n1-112upd:~ # kubectl get pods -n longhorn-system -o custom-columns="Name":metadata.name,"Priority":.spec.priority
      Name                                                  Priority
      backing-image-manager-36c7-86b7                       2000000000
      backing-image-manager-36c7-ecd5                       2000000000
      csi-attacher-6d46d6df-kwnmd                           2000000000
      csi-attacher-6d46d6df-np5fd                           2000000000
      csi-attacher-6d46d6df-svgwf                           2000000000
      csi-provisioner-789fb46d84-2tml2                      2000000000
      csi-provisioner-789fb46d84-9fvm2                      2000000000
      csi-provisioner-789fb46d84-c45tq                      2000000000
      csi-resizer-7b4c499ff9-9wn8l                          2000000000
      csi-resizer-7b4c499ff9-fgbzh                          2000000000
      csi-resizer-7b4c499ff9-llcmj                          2000000000
      csi-snapshotter-865d9b6f88-b8bp2                      2000000000
      csi-snapshotter-865d9b6f88-nm4z6                      2000000000
      csi-snapshotter-865d9b6f88-wdl8k                      2000000000
      engine-image-ei-1d169b76-czfnq                        2000000000
      engine-image-ei-1d169b76-q2pb7                        2000000000
      engine-image-ei-a5371358-ljnct                        2000000000
      engine-image-ei-a5371358-vgj8z                        2000000000
      instance-manager-e-0deab547dac0efa29c085ceb997425aa   2000000000
      instance-manager-e-bff8594c3a68ece10876508b97eb7e30   2000000000
      instance-manager-r-0deab547dac0efa29c085ceb997425aa   2000000000
      instance-manager-r-bff8594c3a68ece10876508b97eb7e30   2000000000
      longhorn-admission-webhook-57cf4f4689-94g9n           0
      longhorn-admission-webhook-57cf4f4689-dsvmw           0
      longhorn-conversion-webhook-678ddcc967-ngjd5          0
      longhorn-conversion-webhook-678ddcc967-vxmcd          0
      longhorn-csi-plugin-2z9xq                             2000000000
      longhorn-csi-plugin-mzd6p                             2000000000
      longhorn-driver-deployer-6b9b5bc787-nnqm7             0
      longhorn-loop-device-cleaner-8g5gf                    0
      longhorn-loop-device-cleaner-dpbk6                    0
      longhorn-manager-pmh5r                                0
      longhorn-manager-vkdjt                                0
      longhorn-recovery-backend-fb89c6ddd-7sv84             0
      longhorn-recovery-backend-fb89c6ddd-jbmnv             0
      longhorn-ui-8669f66db6-489w9                          0
      longhorn-ui-8669f66db6-kdlc5                          0   
              
    </details> 

    - [x] - csi-attacher
    - [x] - csi-provisioner
    - [x] - csi-resizer
    - [x] - csi-snapshotter
    - [x] - engine-image-ei
    - [x] - instance-manager-e
    - [x] - instance-manager-r
    - [x] - longhorn-csi-plugin



- [x] 3. Check VM operation after starting VM 

- [x] 4. Check the Monitoring chart functionality. 

    <details><summary>Test Result $~~$ $\color{green}{\textsf{PASS}}$ </summary>
    
    ![image](https://github.com/harvester/harvester/assets/29251855/55d07750-220f-4a70-a497-32cc3e4d5011)

    </details>

**Comment by TachunLin** [comment]
Verified **fixed** on v1.2.0-rc5. Close this issue.

### Result
Verified all test case scenarios of Longhorn priority class for new install and upgraded cluster in
Tet Plans: https://github.com/harvester/harvester/issues/3756#issuecomment-1667973499

### Test Information
* Test Environment: 3 nodes harvester on bare and kvm machine
* Harvester version: v1.2.0-rc5

### Verify Steps
<details><summary>New Cluster</summary>

1. Create a  new cluster.
2. When the cluster is ready, check Longhorn system-managed components have priority class set.
    - Get the value of the `PriorityClass` `system-cluster-critical`:
      ```bash
      kubectl get priorityclass system-cluster-critical
      ```
      
      Output:
      ```
      NAME                      VALUE        GLOBAL-DEFAULT   AGE
      system-cluster-critical   2000000000   false            15d
    - Get pods' priority in the `longhorn-system` namespace:
    
      ```bash
      kubectl get pods -n longhorn-system -o custom-columns="Name":metadata.name,"Priority":.spec.priority
      ```
    
      And verify all system-managed components' pods have the correct priority. System-managed components include:
      - `csi-attacher`
      - `csi-provisioner`
      - `csi-resizer`
      - `csi-snapshotter`
      - `engine-image-ei`
      - `instance-manager-e`
      - `instance-manager-r`
      - `longhorn-csi-plugin`
</details>

<details><summary>Upgrade path</summary>

Refer to the [document](https://github.com/harvester/harvesterhci.io/blob/fb8a5558f6e89a8ae9ca5f7485ad8b18c7793d81/kb/2023-07-25/configure_priority_class_longhorn.md) to configure Longhorn priority class 
1. Upgrade to v1.2.0-rc5
2. Shutdown all the VMs
```
n1-112upd:~ # kubectl get vmi -A -o json | jq -r '.items[] | [.metadata.name, .metadata.namespace] | @tsv' | while IFS=$'\t' read -r name namespace; do
>       if [ -z "$name" ]; then
>         break
>       fi
>       echo "Stop ${namespace}/${name}"
>       virtctl stop $name -n $namespace
>     done
Stop default/opensuse
VM opensuse was scheduled to stop
Stop default/ubuntu
VM ubuntu was scheduled to stop

n1-112upd:~ # kubectl get vmi -A
No resources found
```

3. Scale down the `Prometheus deployment`
```
n1-112upd:~ # kubectl patch -n cattle-monitoring-system prometheus/rancher-monitoring-prometheus --patch '{"spec": {"replicas": 0}}' --type merge && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system statefulset/prometheus-rancher-monitoring-prometheus
prometheus.monitoring.coreos.com/rancher-monitoring-prometheus patched
statefulset rolling update complete 0 pods at revision prometheus-rancher-monitoring-prometheus-96b9bc787...
``` 

4. Scale down the `AlertManager deployment`

```
n1-112upd:~ # kubectl patch -n cattle-monitoring-system alertmanager/rancher-monitoring-alertmanager --patch '{"spec": {"replicas": 0}}' --type merge && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system statefulset/alertmanager-rancher-monitoring-alertmanager
alertmanager.monitoring.coreos.com/rancher-monitoring-alertmanager patched
statefulset rolling update complete 0 pods at revision alertmanager-rancher-monitoring-alertmanager-866584556d...

```

5. Scale down the Grafana deployment.
```
n1-112upd:~ # kubectl scale --replicas=0 deployment/rancher-monitoring-grafana -n cattle-monitoring-system && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system deployment/rancher-monitoring-grafana
deployment.apps/rancher-monitoring-grafana scaled
deployment "rancher-monitoring-grafana" successfully rolled out
```

6. Scale down vm-import-controller pods
Check if the [vm-import-controller addon](https://docs.harvesterhci.io/v1.1/advanced/vmimport) is enabled and configured with a persistent volume:

```
n1-112upd:~ # kubectl get pvc -n harvester-system harvester-vm-import-controller
Error from server (NotFound): persistentvolumeclaims "harvester-vm-import-controller" not found
```


7. Set the priority-class setting
verify all volumes' STATE are detached:
```
n1-112upd:~ # kubectl get volumes.longhorn.io -A
NAMESPACE         NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE          NODE        AGE
longhorn-system   pvc-09a3cc0e-c459-421b-8a88-9ca0cdc47a36   detached   unknown                  10737418240               15h
longhorn-system   pvc-1420fde1-2f3f-4f21-8f96-d1d30607cb83   detached   unknown                  2147483648                13h
longhorn-system   pvc-1c9659b8-29fb-4c05-ae23-0ac91e1447af   attached   degraded                 1073741824    n1-112upd   15h
longhorn-system   pvc-25e37966-86da-4124-83fc-576fcc0f322b   detached   unknown                  2147483648                17h
longhorn-system   pvc-32ed95f3-3f41-420a-bc15-8db1a49699e5   detached   unknown                  53687091200               17h
longhorn-system   pvc-38073b90-a322-4a3b-b45d-39b55fa20e2d   detached   unknown                  5368709120                17h
longhorn-system   pvc-9fae78f4-bdb7-42b0-8c3f-bd0e59f91434   detached   unknown                  10737418240               15h
```

8. Detach the attached volume or click `Dismiss` on the upgrade dashboard
![image](https://github.com/harvester/harvester/assets/29251855/520695f2-bae6-4473-91c2-a8c1adffc4e3)

```
n1-112upd:~ # kubectl get volumes.longhorn.io -A
NAMESPACE         NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE          NODE   AGE
longhorn-system   pvc-09a3cc0e-c459-421b-8a88-9ca0cdc47a36   detached   unknown                  10737418240          15h
longhorn-system   pvc-1420fde1-2f3f-4f21-8f96-d1d30607cb83   detached   unknown                  2147483648           14h
longhorn-system   pvc-25e37966-86da-4124-83fc-576fcc0f322b   detached   unknown                  2147483648           17h
longhorn-system   pvc-32ed95f3-3f41-420a-bc15-8db1a49699e5   detached   unknown                  53687091200          17h
longhorn-system   pvc-38073b90-a322-4a3b-b45d-39b55fa20e2d   detached   unknown                  5368709120           17h
longhorn-system   pvc-9fae78f4-bdb7-42b0-8c3f-bd0e59f91434   detached   unknown                  10737418240          15h
```

9. Set the `priority-class` setting: 
```
n1-112upd:~ # kubectl patch -n longhorn-system settings.longhorn.io priority-class --patch '{"value": "system-cluster-critical"}' --type merge
setting.longhorn.io/priority-class patched
```

10. Get the value of the PriorityClass system-cluster-critical:
```
n1-112upd:~ # kubectl get priorityclass system-cluster-critical
NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            17h
```

11. Get pods' priority in the longhorn-system namespace:
```
n1-112upd:~ # kubectl get pods -n longhorn-system -o custom-columns="Name":metadata.name,"Priority":.spec.priority
Name                                                  Priority
backing-image-manager-36c7-86b7                       2000000000
backing-image-manager-36c7-ecd5                       2000000000
csi-attacher-6d46d6df-kwnmd                           2000000000
csi-attacher-6d46d6df-np5fd                           2000000000
csi-attacher-6d46d6df-svgwf                           2000000000
csi-provisioner-789fb46d84-2tml2                      2000000000
csi-provisioner-789fb46d84-9fvm2                      2000000000
csi-provisioner-789fb46d84-c45tq                      2000000000
csi-resizer-7b4c499ff9-9wn8l                          2000000000
csi-resizer-7b4c499ff9-fgbzh                          2000000000
csi-resizer-7b4c499ff9-llcmj                          2000000000
csi-snapshotter-865d9b6f88-b8bp2                      2000000000
csi-snapshotter-865d9b6f88-nm4z6                      2000000000
csi-snapshotter-865d9b6f88-wdl8k                      2000000000
engine-image-ei-1d169b76-czfnq                        2000000000
engine-image-ei-1d169b76-q2pb7                        2000000000
engine-image-ei-a5371358-ljnct                        2000000000
engine-image-ei-a5371358-vgj8z                        2000000000
instance-manager-e-0deab547dac0efa29c085ceb997425aa   2000000000
instance-manager-e-bff8594c3a68ece10876508b97eb7e30   2000000000
instance-manager-r-0deab547dac0efa29c085ceb997425aa   2000000000
instance-manager-r-bff8594c3a68ece10876508b97eb7e30   2000000000
longhorn-admission-webhook-57cf4f4689-94g9n           0
longhorn-admission-webhook-57cf4f4689-dsvmw           0
longhorn-conversion-webhook-678ddcc967-ngjd5          0
longhorn-conversion-webhook-678ddcc967-vxmcd          0
longhorn-csi-plugin-2z9xq                             2000000000
longhorn-csi-plugin-mzd6p                             2000000000
longhorn-driver-deployer-6b9b5bc787-nnqm7             0
longhorn-loop-device-cleaner-8g5gf                    0
longhorn-loop-device-cleaner-dpbk6                    0
longhorn-manager-pmh5r                                0
longhorn-manager-vkdjt                                0
longhorn-recovery-backend-fb89c6ddd-7sv84             0
longhorn-recovery-backend-fb89c6ddd-jbmnv             0
longhorn-ui-8669f66db6-489w9                          0
longhorn-ui-8669f66db6-kdlc5                          0

```

12. Scale up vm-import-controller pods
```
n1-112upd:~ # kubectl scale --replicas=1 deployment/harvester-vm-import-controller -n harvester-system && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n harvester-system deployment/harvester-vm-import-controller
Error from server (NotFound): deployments.apps "harvester-vm-import-controller" not found
```

13. Scale up the Prometheus deployment.

```
n1-112upd:~ # kubectl patch -n cattle-monitoring-system prometheus/rancher-monitoring-prometheus --patch '{"spec": {"replicas": 1}}' --type merge && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system statefulset/prometheus-rancher-monitoring-prometheus
prometheus.monitoring.coreos.com/rancher-monitoring-prometheus patched
Waiting for 1 pods to be ready...
Waiting for 1 pods to be ready...
statefulset rolling update complete 1 pods at revision prometheus-rancher-monitoring-prometheus-96b9bc787...

```

14. Scale up the AlertManager deployment.
```
n1-112upd:~ # kubectl patch -n cattle-monitoring-system alertmanager/rancher-monitoring-alertmanager --patch '{"spec": {"replicas": 1}}' --type merge && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system statefulset/alertmanager-rancher-monitoring-alertmanager
alertmanager.monitoring.coreos.com/rancher-monitoring-alertmanager patched
Waiting for 1 pods to be ready...
statefulset rolling update complete 1 pods at revision alertmanager-rancher-monitoring-alertmanager-575b966c5d...
```

15. Scale down the Grafana deployment.
```
n1-112upd:~ # kubectl scale --replicas=1 deployment/rancher-monitoring-grafana -n cattle-monitoring-system && \
>     sleep 5 && \
>     kubectl rollout status --watch=true -n cattle-monitoring-system deployment/rancher-monitoring-grafana
deployment.apps/rancher-monitoring-grafana scaled
Waiting for deployment "rancher-monitoring-grafana" rollout to finish: 0 of 1 updated replicas are available...
deployment "rancher-monitoring-grafana" successfully rolled out
```

16. Start virtual machines
```
n1-112upd:~ # kubectl get vms -A
NAMESPACE   NAME       AGE   STATUS    READY
default     opensuse   16h   Stopped   False
default     ubuntu     16h   Stopped   False
n1-112upd:~ # virtctl start opensuse -n default
VM opensuse was scheduled to start
n1-112upd:~ # virtctl start ubuntu -n default
VM ubuntu was scheduled to start
```

</details>
